### PR TITLE
inbound stats

### PR DESCRIPTION
**Pull request by zhuker**
*Issue #, if available:*

this PR implements RTC inbound stats as defined here https://www.w3.org/TR/webrtc-stats/#dom-rtcinboundrtpstreamstats

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by zhuker** [comment]
just a placeholder for now

**Comment by zhuker** [comment]
@MushMal our receiver code needs more love to implements stats properly

**Comment by disa6302** [comment]
@zhuker , 

Apologies for the delay. I will spend some time today to review your PR.

File: src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h [hunk 1]
```diff
@@ -320,6 +328,150 @@ typedef struct {
...
+} RtcReceivedRtpStreamStats, *PRtcReceivedRtpStreamStats;
+
+/**
+ * @brief The RTCInboundRtpStreamStats dictionary represents the measurement metrics for the incoming RTP media stream. The timestamp reported in the
+ * statistics object is the time at which the data was sampled.
+ *
+ * Reference: https://www.w3.org/TR/webrtc-stats/#dom-rtcinboundrtpstreamstats
+ */
+typedef struct {
+    RtcReceivedRtpStreamStats received; // dictionary RTCInboundRtpStreamStats : RTCReceivedRtpStreamStats
```
Comment (disa6302) [inline src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h:373]: These do not seem to be part of inbound stats per spec. Can we add a comment to ensure we capture why we add RtcReceivedRtpStreamStats as part of inbound stats structure here?

File: src/source/PeerConnection/JitterBuffer.h [hunk 2]
```diff
@@ -20,6 +20,12 @@ typedef struct {
     FrameDroppedFunc onFrameDroppedFn;
     DepayRtpPayloadFunc depayPayloadFn;
 
+    // used for calculating interrival jitter https://tools.ietf.org/html/rfc3550#section-6.4.1
```
Comment (disa6302) [inline src/source/PeerConnection/JitterBuffer.h:23]: Typo: `interrival ` -> `interarrival`

File: src/source/PeerConnection/PeerConnection.c [hunk 3]
```diff
@@ -179,10 +177,36 @@ STATUS sendPacketToRtpReceiver(PKvsPeerConnection pKvsPeerConnection, PBYTE pBuf
...
+
+            // https://tools.ietf.org/html/rfc3550#section-6.4.1
+            // https://tools.ietf.org/html/rfc3550#appendix-A.8
+            // interarrival jitter
+            // arrival, the current time in the same units.
+            // r_ts, the timestamp from   the incoming packet
+            arrival = KVS_CONVERT_TIMESCALE(GETTIME(), HUNDREDS_OF_NANOS_IN_A_SECOND, pTransceiver->pJitterBuffer->clockRate);
+            r_ts = pRtpPacket->header.timestamp;
+            transit = arrival - r_ts;
+            d = transit - pTransceiver->pJitterBuffer->transit;
```
Comment (disa6302) [inline src/source/PeerConnection/PeerConnection.c:200]: Any possibility `d` evaluates to a negative number?

File: src/source/PeerConnection/Rtp.c [hunk 4]
```diff
@@ -136,27 +136,43 @@ STATUS transceiverOnBandwidthEstimation(PRtcRtpTransceiver pRtcRtpTransceiver, U
     return retStatus;
 }
 
+STATUS transceiverOnPictureLoss(PRtcRtpTransceiver pRtcRtpTransceiver, UINT64 customData, RtcOnPictureLoss onPictureLoss)
```
Comment (disa6302) [inline src/source/PeerConnection/Rtp.c:139]: How is this related to the PR?

File: src/source/PeerConnection/Rtp.c [hunk 5]
```diff
@@ -319,35 +335,35 @@ STATUS writeFrame(PRtcRtpTransceiver pRtcRtpTransceiver, PFrame pFrame)
...
-        pKvsRtpTransceiver->sender.outboundStats.framesPerSecond = fps;
+        pKvsRtpTransceiver->outboundStats.framesPerSecond = fps;
     }
     pKvsRtpTransceiver->sender.lastKnownFrameCountTime = now;
-    pKvsRtpTransceiver->sender.lastKnownFrameCount = pKvsRtpTransceiver->sender.outboundStats.framesEncoded;
-    pKvsRtpTransceiver->sender.outboundStats.sent.bytesSent += bytesSent;
-    pKvsRtpTransceiver->sender.outboundStats.sent.packetsSent++;
+    pKvsRtpTransceiver->sender.lastKnownFrameCount = pKvsRtpTransceiver->outboundStats.framesEncoded;
+    pKvsRtpTransceiver->outboundStats.sent.bytesSent += bytesSent;
+    pKvsRtpTransceiver->outboundStats.sent.packetsSent++;
```
Comment (disa6302) [inline src/source/PeerConnection/Rtp.c:348]: Shouldn't this be `pKvsRtpTransceiver->outboundStats.sent.packetsSent += packetsSent`?

File: src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h [hunk 6]
```diff
@@ -320,6 +328,150 @@ typedef struct {
...
+                           //!< and is a cumulative metric as defined in Appendix A (i) of [RFC7004].
+} RtcReceivedRtpStreamStats, *PRtcReceivedRtpStreamStats;
+
+/**
+ * @brief The RTCInboundRtpStreamStats dictionary represents the measurement metrics for the incoming RTP media stream. The timestamp reported in the
+ * statistics object is the time at which the data was sampled.
+ *
+ * Reference: https://www.w3.org/TR/webrtc-stats/#dom-rtcinboundrtpstreamstats
+ */
+typedef struct {
```
Comment (disa6302) [inline src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h:372]: Can we add TODOs to the ones that are not implemented yet?

File: src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h [hunk 7]
```diff
@@ -320,6 +328,150 @@ typedef struct {
     UINT64 roundTripTimeMeasurements; //!< Total number of RTCP RR blocks received for this SSRC that contain a valid round trip time
 } RtcRemoteInboundRtpStreamStats, *PRtcRemoteInboundRtpStreamStats;
 
+typedef struct {
```
Comment (disa6302) [inline src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h:331]: Can we add TODOs to the ones that are not implemented?

File: tst/JitterBufferFunctionalityTest.cpp [hunk 8]
```diff
@@ -71,7 +71,7 @@ TEST_F(JitterBufferFunctionalityTest, continousPacketsComeInOrder)
     setPayloadToFree();
```
Comment (disa6302) [inline tst/JitterBufferFunctionalityTest.cpp:71]: Could you also add an inbound RTP stats API test in MetricsApiTest.cpp if it is possible to instrument a testing of the public API?

File: src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h (hunk 1 above)
Comment (zhuker) [reply src/include/com/amazonaws/kinesis/video/webrtcclient/Stats.h:373]: image

File: src/source/PeerConnection/Rtp.c (hunk 4 above)
Comment (zhuker) [reply src/source/PeerConnection/Rtp.c:139]: i refactored handling of PLI packets (`transceiverOnPictureLoss` is public API for PLI callback)
to be absolutely honest it is not really related to this PR :) 
just added for completeness and because i needed it in production

File: src/source/PeerConnection/PeerConnection.c (hunk 3 above)
Comment (zhuker) [reply src/source/PeerConnection/PeerConnection.c:200]: let me doublecheck

File: src/source/PeerConnection/Rtp.c (hunk 4 above)
Comment (zhuker) [reply src/source/PeerConnection/Rtp.c:139]: want me to take it out?

Comment (disa6302) [reply src/source/PeerConnection/Rtp.c:139]: Cool! It would be super helpful if you could create a new PR with the onPictureLoss changes. It is just to make sure that related changes go into a single commit at the end so that it is independently tested. Thats all :)

File: src/source/PeerConnection/PeerConnection.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhuker) [reply src/source/PeerConnection/PeerConnection.c:200]: delta?

File: src/source/PeerConnection/PeerConnection.c [hunk 9]
```diff
@@ -179,10 +177,36 @@ STATUS sendPacketToRtpReceiver(PKvsPeerConnection pKvsPeerConnection, PBYTE pBuf
...
+            pTransceiver->pJitterBuffer->transit = transit;
+            pTransceiver->pJitterBuffer->jitter += (1. / 16.) * ((DOUBLE) ABS(d) - pTransceiver->pJitterBuffer->jitter);
```
[the remark this answers is not in the document]
Comment (zhuker) [reply src/source/PeerConnection/PeerConnection.c:202]: whats wrong with it being zero?

[the remark this answers is not in the document]
Comment (MushMal) [reply src/source/PeerConnection/PeerConnection.c]: D'oh, my fault. Nothing - I interpreted the * as /

BTW, do we have our older GCC not complain about 1. and 16. literals? Don't recall the spec but shouldn't it be 1.0?

[the remark this answers is not in the document]
Comment (zhuker) [reply src/source/PeerConnection/PeerConnection.c]: CI seems to be happy about it

File: src/source/PeerConnection/Rtp.c (hunk 4 above)
Comment (zhuker) [reply src/source/PeerConnection/Rtp.c:139]: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/692

**Comment by zhuker** [comment]
@disa6302 separate `onPictureLoss` PR https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/692

File: src/source/PeerConnection/PeerConnection.c (hunk 3 above)
Comment (zhuker) [reply src/source/PeerConnection/PeerConnection.c]: changed to signed

**Comment by zhuker** [comment]
@disa6302 @MushMal i think this PR is ready for merge
anything outstanding?

**Comment by MushMal** [comment]
> @disa6302 @MushMal i think this PR is ready for merge
> anything outstanding?

Looks good. @disa6302 ?

File: src/source/PeerConnection/SessionDescription.c [hunk 10]
```diff
@@ -875,6 +875,11 @@ STATUS setReceiversSsrc(PSessionDescription pRemoteSessionDescription, PDoubleLi
                         ((isVideoCodec && isVideoMediaSection) || (isAudioCodec && isAudioMediaSection))) {
                         // Finish iteration, we assigned the ssrc move on to next media section
                         pKvsRtpTransceiver->jitterBufferSsrc = ssrc;
+                        pKvsRtpTransceiver->inboundStats.received.rtpStream.ssrc = ssrc;
+                        STRNCPY(pKvsRtpTransceiver->inboundStats.received.rtpStream.kind,
+                                pKvsRtpTransceiver->transceiver.receiver.track.kind == MEDIA_STREAM_TRACK_KIND_VIDEO ? "video" : "audio",
+                                sizeof(DOMString));
```
Comment (disa6302) [inline src/source/PeerConnection/SessionDescription.c:881]: Use `ARRAY_SIZE(DOMString)` instead of SIZEOF.

**Comment by zhuker** [comment]
@disa6302 :tada: i finally learned how to squash commits into one
